File: Algo/LeadToOffer/75_CutRope.cpp
// dp
class Solution {
public:
    int cuttingRope(int n) {
        if(n < 4) return n - 1;
        vector<int> dp(n+1);
        dp[0] = 0;
        dp[1] = 1;
        dp[2] = 2;
        dp[3] = 3;

        for(int i=4; i<n+1; i++){
            dp[i] = 0;
            for(int k=1; k<=i / 2; k++){ // At least 1. At most k/2 since symmetric.
                dp[i] = max(dp[i], dp[k]*dp[i-k]);
            }
        }
        return dp[n];
    }
};

File: Algo/LeadToOffer/75_CutRope.py
class Solution(object):
    def cuttingRope(self, n):
        """
        :type n: int
        :rtype: int
        """
        if n == 1:
            return 1
        if n == 2:
            return 1
        if n == 3:
            return 2

        dp = [0] * (n+1)
        dp[1] = 1
        dp[2] = 2       # max single 2
        dp[3] = 3       # max single 3

        for i in range(4, n+1):
            dp[i] = max([dp[i-k]*dp[k] for k in range(1, i/2+1)])

        return dp[-1]